hazelcast/src/main/java/com/hazelcast/internal/partition/operation/MigrationOperation.java

```diff
@@ -221,44 +168,39 @@ private void onOperationFailure(Operation op, Throwable e) {
     }
 
     @Override
-    protected void writeInternal(ObjectDataOutput out) throws IOException {
-        super.writeInternal(out);
-        out.writeLongArray(replicaVersions);
-        int size = tasks != null ? tasks.size() : 0;
-        out.writeInt(size);
-        if (size > 0) {
-            for (Operation task : tasks) {
-                out.writeObject(task);
-            }
+    public int getId() {
+        return PartitionDataSerializerHook.MIGRATION;
+    }
+
+    @Override
+    void onMigrationStart() {
+        if (firstFragment) {
+            super.onMigrationStart();
         }
     }
 
     @Override
-    protected void readInternal(ObjectDataInput in) throws IOException {
-        super.readInternal(in);
-        replicaVersions = in.readLongArray();
-        int size = in.readInt();
-        if (size > 0) {
-            tasks = new ArrayList<Operation>(size);
-            for (int i = 0; i < size; i++) {
-                Operation op = in.readObject();
-                tasks.add(op);
-            }
-        } else {
-            tasks = Collections.emptyList();
+    void onMigrationComplete(boolean result) {
+        if (lastFragment) {
+            super.onMigrationComplete(result);
         }
     }
 
     @Override
-    protected void toString(StringBuilder sb) {
-        int numberOfTasks = tasks == null ? 0 : tasks.size();
-        sb.append(", migration=").append(migrationInfo);
-        sb.append(", replicaVersions=").append(Arrays.toString(replicaVersions));
-        sb.append(", numberOfTasks=").append(numberOfTasks);
+    protected void writeInternal(ObjectDataOutput out) throws IOException {
+        super.writeInternal(out);
+        fragmentMigrationState.writeData(out);
+        out.writeBoolean(firstFragment);
+        out.writeBoolean(lastFragment);
     }
 
     @Override
-    public int getId() {
-        return PartitionDataSerializerHook.MIGRATION;
+    protected void readInternal(ObjectDataInput in) throws IOException {
+        super.readInternal(in);
+        fragmentMigrationState = new ReplicaFragmentMigrationState();
```

Comment:
Not necessarily better but it will be consistent with other IDS deserialisations.